expertise/service/README.md

```diff
@@ -0,0 +1,145 @@
+```
+
+By default, the app will run on `http://localhost:5000`. The endpoint `/expertise/test` should show a simple page indicating that Flask is running.
+
+## Accessing the Endpoints
+Access to all endpoints requires authentication in the header of the request.
+
+**`POST /expertise`**
+
+**Request Body:**
+
+```
+{
+  config: JSON object containing options for both create_dataset and run_expertise
+}
+```
+
+**Returns:**
+
+`{ user_id: str, job_id: int }`
+
+---
+
+**`GET /jobs`**
+
+**Request Body:**
+
+`{}`
+
+**Returns:**
+
+```
+{
+  'dataset': [
+    {
+      job_name: str,
+      job_id: int,
+      status: str,
+    }, ...
+  ],
+  'expertise': [
+    {
+      job_name: str,
+      job_id: int,
+      status: str,
+    }, ...
+  ]
+}
+```
+
+---
+
+**`GET /results`**
+
+**Request Body:**
+
+`{ job_id: int, delete_on_get: boolean }`
+
+**Returns:**
+
+```
+{
+  results: [
+    {
+      submission: str,
+      user: str,
+      status: float,
+    }, ...
+  ]
+}
+```
+
+# Basic Queue Structures
+## JobQueue
+
+JobData is a dataclass that's used to store metadata about each jobs and it is defined with the following fields (from `queue.py`):
+```
+    id: str = field(
+        metadata={"help": "The profile id at the time of submission"},
+    )
+    job_name: str = field(
+        metadata={"help": "The name of the job specified in the submitted config file"},
+    )
+    config: dict = field(
+        metadata={"help": "The submitted configuration file as a dictionary"},
+    )
+    timeout: int = field(
+        default=0,
+        metadata={"help": "The maximum amount of time to run this job"},
+    )
+```
+
+JobQueue is an abstract wrapper for the built-in Python `queue.Queue()` class. The built-in queue offers a standard FIFO queue that allows both asynchronous and blocking calls for `put()` and `get()`. JobQueue works under the assumption that each job accepts a configuration file and makes modifications to the filesystem to store intermediate and final results. JobQueue offers a set of services on top the built-in queue including:
+1.  Keeping track of a history of submitted jobs
+2.  Pre-emptive canceling of jobs that are still in queue
+3.  Timing out of possibly long running jobs
+4.  Retrieving the current status of an arbitrary job
+5.  Multi-threaded multi-process handling of submitted jobs
+6.  Allocating a directory on a per-job basis to enforce job isolation
+
+JobQueue defines several status codes that are free to asynchronously query during a job's lifetime:
+- `queued` -- The job is currently awaiting processing by a worker
+- `processing` -- The job is currently being worked on by a worker
+- `completed` -- The job has finished and the results are currently stored on the server
+- `stale` -- The job has been canceled before it arrived at processing
+- `timeout` -- The job has exceeded the specified timeout (there is no timeout by default)
+- `error` -- While processing, the job process has encountered an error
```

Comment:
This is a very impressive queue! I'm wondering if this should be in the definition of the class JobQueue instead of the README.

Reply:
I'm also impressed about the implementation!!!

I see the TwoStepQueue implementation is a little bit complex and difficult to follow. To make the things easier, I wouldn't worry about the token expired error and implement one single queue with no inner queues. 

This queue will have one single job that runs two tasks:

- execute_create_dataset
- execute_expertise

If then token is expired then we set a status error and the user should try again with a refreshed token.

Another option to avoid this error is to support username/password in the request headers and the API can instantiate the `openreview.Client` with a username and password instead of a token. 

Have you ever checked the Celery library to work with queues? https://docs.celeryproject.org/en/stable/getting-started/first-steps-with-celery.html#redis

I think this library can save you a lot of work. 

This is another option: https://python-rq.org/ but @purujitgoyal said that he had some issues with this one. 

I found another one: https://github.com/coleifer/huey